### PR TITLE
Revised LinkGenerator

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs [hunk 1]
```diff
@@ -169,24 +180,24 @@ internal sealed class DefaultLinkGenerator : LinkGenerator
...
+                ambientValues: null,
+                scheme: scheme,
+                host: host,
+                pathBase: pathBase,
+                fragment: fragment,
+                options: options);
         }
 
-        public override LinkGenerationTemplate GetTemplateByAddress<TAddress>(TAddress address)
+        public override LinkGenerationTemplate GetTemplateByAddress<TAddress>(TAddress address, LinkGenerationTemplateOptions options = default)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs]: What are the rules on default vs null? Elseware we use null, e.g. LinkOptions options = null.

File: test/Microsoft.AspNetCore.Routing.Tests/LinkGeneratorRouteValuesAddressExtensionsTest.cs [hunk 2]
```diff
@@ -16,6 +17,42 @@ namespace Microsoft.AspNetCore.Routing
...
+                "Home/Index/{id}",
+                defaults: new { controller = "Home", action = "Index", },
+                metadata: new[] { new RouteValuesAddressMetadata(routeName: null, new RouteValueDictionary(new { controller = "Home", action = "Index", })) });
+            var endpoint2 = EndpointFactory.CreateRouteEndpoint(
+                "Home/Index/{id?}",
+                defaults: new { controller = "Home", action = "Index", },
+                metadata: new[] { new RouteValuesAddressMetadata(routeName: null, new RouteValueDictionary(new { controller = "Home", action = "Index", })) });
+
+            var linkGenerator = CreateLinkGenerator(endpoint1, endpoint2);
+
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Routing.Tests/LinkGeneratorRouteValuesAddressExtensionsTest.cs]: nit: too liney

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs [hunk 3]
```diff
@@ -454,17 +458,73 @@ public void GetUriByAddress_WithHttpContext_IncludesAmbientValues()
...
+            // Arrange
+            var endpoint1 = EndpointFactory.CreateRouteEndpoint("{controller}/{action}/{id}", metadata: new object[] { new IntMetadata(1), });
+            var endpoint2 = EndpointFactory.CreateRouteEndpoint("{controller}/{action}/{id?}", metadata: new object[] { new IntMetadata(1), });
+
+            var linkGenerator = CreateLinkGenerator(endpoint1, endpoint2);
+
+            var httpContext = CreateHttpContext();
+            httpContext.Request.Scheme = "http";
+            httpContext.Request.Host = new HostString("example.com");
+            httpContext.Request.PathBase = "/Foo";
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Do schema and host have any effect or are they just required for the test to pass?

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGenerationTemplateTest.cs [hunk 4]
```diff
@@ -130,14 +133,42 @@ public void GetPath_WithHttpContext_IncludesAmbientValues()
...
+            httpContext.Request.Host = new HostString("example.com");
+
+            // Act
+            var uri = template.GetPath(httpContext, values: new RouteValueDictionary(new { action = "Index", }));
+
+            // Assert
+            Assert.Null(uri);
+        }
+
+
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGenerationTemplateTest.cs]: nit: too liney

File: src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs]: default means null for reference types. I just make this code self-consistent.

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs (hunk 3 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Yeah, bad copy paste they don't do anything here. Removed.

**Comment by rynowak** [comment]
Building MVC against this for safety. I'm not sure if I'll need a feature branch or not.

**Comment by rynowak** [comment]
It does 😢

**Pull request merged**